src/test/groovy/com/rabbitmq/http/client/ClientSpec.groovy

```diff
@@ -1709,6 +1709,130 @@ class ClientSpec extends Specification {
+
+    then: "list of upstreams that contains the new upstream is returned"
+    verifyUpstreamDefinitions(vhost, upstreams, upstreamName)
+
+    cleanup:
+    client.deleteUpstream(vhost, upstreamName)
+  }
+
+  def "GET /api/parameters/federation-upstream declare and get at non-root vhost"() {
+    given: "an upstream"
+    final vhost = "foo"
+    final upstreamName = "upstream2"
+    client.createVhost(vhost)
+    declareUpstream(client, vhost, upstreamName)
+
+    when: "client requests the upstreams"
+    final upstreams = awaitEventPropagation { client.getUpstreams(vhost) }
+
+    then: "list of upstreams that contains the new upstream is returned"
+    verifyUpstreamDefinitions(vhost, upstreams, upstreamName)
+
+    cleanup:
+    client.deleteUpstream(vhost,upstreamName)
+    client.deleteVhost(vhost)
+  }
+
+  def "PUT /api/parameters/federation-upstream with null upstream uri"() {
+    given: "an Upstream without upstream uri"
+    UpstreamDetails upstreamDetails = new UpstreamDetails()
+
+    when: "client tries to declare an Upstream"
+    client.declareUpstream("/", "upstream3", upstreamDetails)
+
+    then: "an illegal argument exception is thrown"
+    thrown(IllegalArgumentException)
+  }
+
+  def "DELETE /api/parameters/federation-upstream/{vhost}/{name}"() {
+    given: "upstream upstream4 in vhost /"
+    final vhost = "/"
+    final upstreamName = "upstream4"
+    declareUpstream(client, vhost, upstreamName)
+
+    List<ParameterWrapper<UpstreamDetails>> upstreams = awaitEventPropagation { client.getUpstreams() }
+    verifyUpstreamDefinitions(vhost, upstreams, upstreamName)
+
+    when: "client deletes upstream upstream4 in vhost /"
+    client.deleteUpstream(vhost, upstreamName)
+
+    and: "upstream list in / is reloaded"
+    upstreams = client.getUpstreams()
+
+    then: "upstream4 no longer exists"
+    upstreams.find { it.name.equals(upstreamName) } == null
+  }
+
+  def "GET /api/parameters/federation-upstream-set declare and get"() {
+    given: "an upstream set with two upstreams"
+    final vhost = "/"
+    final upstreamSetName = "upstream-set-1"
+    final upstreamA = "A"
+    final upstreamB = "B"
+    declareUpstream(client, vhost, upstreamA);
+    declareUpstream(client, vhost, upstreamB);
+    final d1 = new UpstreamSetDetails()
+    d1.setUpstream(upstreamA)
+    d1.setExchange("exchangeA")
+    final d2 = new UpstreamSetDetails()
+    d2.setUpstream(upstreamB)
+    d2.setExchange("exchangeB")
+    final detailsSet = new ArrayList()
+    detailsSet.add(d1)
+    detailsSet.add(d2)
+    client.declareUpstreamSet(vhost, upstreamSetName, detailsSet)
+
+    when: "client requests the upstream set list"
+    final upstreamSets = awaitEventPropagation { client.getUpstreamSets() }
+
+    then: "upstream set with two upstreams is returned"
+    assert !upstreamSets.isEmpty()
+    ParameterWrapper<UpstreamSetDetails> upstreamSet = upstreamSets.find { it.name.equals(upstreamSetName) }
+    assert upstreamSet != null
```

Comment:
Not sure the `assert`s are necessary here.

Reply:
You are right, `assert`s are not necessary.